acstools/acsphotcte.py

```diff
@@ -0,0 +1,289 @@
+#!/usr/bin/env python
+"""
+The ACS Photometric CTE API is a programmatic interface for the
+`ACS Photometric CTE Webtool <https://acsphotometriccte.stsci.edu>`_.
+The API is a cloud-based service that employs a serverless approach on AWS
+with API Gateway and Lambda to compute the photometric CTE corrections
+using the model described in ACS ISR 2012-02.
+The model corrects ACS/WFC aperture photometry extracted from FLT images for
+CTE losses using the most recent formula from Chiaberge's ISR 2012-05. It is
+only calibrated for photometry obtained after SM4 in May 2009. For pre-SM4
+data, please see Chiaberge et al. 2009 (ISR 2009-01), or use pixel-based
+CTE-corrected files obtained from MAST. Currently,  only 3 and 5
+pixel aperture radii are supported. The model is designed to compute the
+CTE losses as a function of the number of Y transfers, sky background,
+source flux, and observation date.
+
+
+Usage Guidelines
+----------------
+As this service is hosted on AWS, there are some guidelines that should be
+followed when utilizing this tool. For each call to
+:py:meth:`PhotCTEAPI.correct_photomerty`, AWS will assemble the compute resources,
+install the software, compute the CTE corrections, send the result back,
+and then terminate all resources. Hence, users should try to process as many
+sources as they can in a single function call. Testing has shown that the
+optimal number is < 25000, but the service can handle as many as 150,000
+sources in a single request.
+
+
+Example
+-------
+Get the CTE corrected FLT photometry for a list of 1000 sources:
+
+>>> import numpy as np
+>>> from acstools import acsphotcte
+>>> ytransfers = 2048 * np.random.random(size=1000)
+>>> fluxes = 20000*np.random.random(size=len(ytransfers))
+>>> local_skys = 80*np.random.random(size=len(ytransfers))
+>>> mjd = 59341.
+>>> radius = 3.
+>>> cte_corrections = acsphotcte.correct_photometry(
+...    radius=radius,
+...    ytransfers=ytransfers,
+...    mjd=mjd,
+...    local_skys=local_skys,
+...    fluxes=fluxes
+...)
+>>> print(cte_corrections)
+"""
+from collections import Iterable
+import json
+import logging
+
+import numpy as np
+import requests
```

Comment:
`requests` should be an optional dependency. It is not used anywhere else in this package (that I can find).

Reply:
Good catch! Haven't had to add optional dependencies before, so I want to double it's through the `extras_require` keyword in `setup()` right? So something like,
```python
install_requires=[
        'astropy>=2',
        'numpy',
        'beautifulsoup4'
    ],
extras_require={'request':'requests'},
ests_require=['pytest'],
```

Reply:
Not only that; It should only be imported inside your method that uses it, not at the module level.

Reply:
Ok, my only question is why not add it to requirements? If we want the module to be useful, then people need to have requests installed or nothing will happen. We had a similar scenario with BeautifulSoup and the `acszpt.py` module. No other module in `acstools` uses BeautifulSoup by `acszpt.py` is useless without that package installed.

Reply:
Good point about the `bs4` requirement. I guess we already set a precedent. If the list keeps growing though, I need to rethink this as not to bloat up the compulsory dependencies of `acstools`.

Reply:
In the future if it's necessary, we could replace the `requests` module with `urllib.request`. For now, I've added `requests` to the dependencies.